### ISSUE TITLE
Create serialized format API for importing / exporting the Puppetfile

### ISSUE THREAD

**Issue by logicminds** [puppetlabs/r10k#710]
In reference to #564.

Many tools, and external processes often need to interface with the Puppetfile in a programmatic form. The puppetfile allows us to interact by evaling the Puppetfile but we also want the ability to interact with r10k via different Puppetfile formats.  Since these formats and serialization methods can differ we need to create an API that will allows users to create their own format or use any 3rd party format easily. 

The API should should meet the following expectations.

1. allow the user to choose which serialization to use (json, yaml, hocon, toml, ...)
2. allow the user to choose which data format to use (the one in #564, direct (@modules.to_json), some_other_format)
3. The API should follow the adapter pattern and provide a contract with only a few methods the user needs to override to create a new format.  This should be simple to create with only a type, serialize, and deserialize method. 
4. The r10k interface should add a new action called export which would allow the user to export the puppetfile to a different format.
5. The r10k interface should add a new action called import which would convert a format to a Puppetfile
6. The r10k interface should also automatically try to import a format given a serialized formatted file.  
7.  A serialized format should include the format type in the data it contains so that r10k can decipher which format it should use to deserialize the data. 
8. R10k should respect environments when exporting or importing files.
9. R10k should automatically detect the type of serialization to use given the extension name of the file. 
10. R10k should import the user's specified format via  settings
11. R10k should import the Puppetfile format via some predefined order if none is set in settings

Possible workflow:

1. r10k --export battletech.json --environment=dev
2. r10k deploy dev   (assumes battletech.json is in r10k-control)
3. r10k --import battletech.json deploy dev  (overrides any setting of format to search for)

Possible API.  Never tested.  Take with a grain of salt.

```
module R10k
  module SerializedFormat
    class BattleTech
         include SerializedFormatAPI

        def serialize(modues)
          modules.to_json
        end

        def deserialize(formatted_object)
           # reconstruct puppetfile ruby object
           to_puppetfile_obj(formatted_object)
        end

        def to_puppetfile_obj(data)
           
        end

        def type
           :battletech
        end

    end
  end
end



```

**Comment by gregswift** [comment]
This is an interesting approach at totally revamping how r10k operates, and thats not a horrible idea. It does seem like a bit significant of a change for just trying to provide more direct support for automation tools like #564 is proposing.

**Comment by reidmv** [comment]
The _most interesting_ piece of this to me is only alluded to in one requirement and in the possible workflow given.

* R10k should respect environments when exporting or importing files.

1. `r10k --export battletech.json --environment=dev`
2. `r10k deploy dev` (assumes battletech.json is in r10k-control)
3. `r10k --import battletech.json deploy dev` (overrides any setting of format to search for)

I have a neutral opinion on natively supporting serialized data Puppetfile formats. I see the utility for machine interaction, but share @logicminds's originally expressed concerns around providing too many options for users and not being simple and prescriptive. I'm therefore not convinced the use case of easing machine interaction merits the change by itself.

That said: I see supporting a serialized format as a possible preparatory step towards addressing a bigger problem. r10k was built as, and excels when being used for, a _development workflow_. Over the last few years I've come around to the perspective though that r10k's use of git branches to represent named environments up to and including production is harmful. I won't go into too much detail here but r10k is not a great _deployment workflow_ for many organizations, especially larger ones.

In the requirement I highlighted above and in the possible workflow, no mention is made of the relationship between environments, git branches, and json files. Is this an implication that part of the proposal changes those relationships in some way?

**tl;dr**: I'm not convinced the use case given merits adding another way to write Puppetfiles. Providing too many options without strong prescription is generally detrimental to usability. Because I care about a different problem for which serial data formats would be beneficial, overall I have a neutral opinion on adding such support.
